src/tools/emscripten.jam

```diff
@@ -72,15 +98,15 @@ toolset.flags emscripten.compile OPTIONS <inlining>on : -Wno-inline ;
 toolset.flags emscripten.compile OPTIONS <inlining>full : -Wno-inline ;
 
 toolset.flags emscripten OPTIONS <debug-symbols>off : -g0 ;
-toolset.flags emscripten OPTIONS <debug-symbols>on : -g4 -s DEMANGLE_SUPPORT=1 ;
+# Currently the highest supported level is -g3. DEMANGLE_SUPPORT, which is, BTW, not supported for JS
+# code generator, should be provided seperately along with --source-map-base when it's needed.
+toolset.flags emscripten OPTIONS <debug-symbols>on : -g3 ;
 toolset.flags emscripten OPTIONS <rtti>off : -fno-rtti ;
 
 toolset.flags emscripten.link OPTIONS <embind>on : --bind ;
 toolset.flags emscripten.link OPTIONS <closure>on : --closure 1 ;
 toolset.flags emscripten.link OPTIONS <closure>full : --closure 2 ;
-toolset.flags emscripten.link OPTIONS <link-optimization>off : --llvm-lto 0 ;
-toolset.flags emscripten.link OPTIONS <link-optimization>on : --llvm-lto 1 ;
-toolset.flags emscripten.link OPTIONS <link-optimization>full : --llvm-lto 3 ;
+toolset.flags emscripten.link OPTIONS <link-optimization>on : -flto ;
```

Comment:
FIY: `link-optimization` option is undocumented and presented only in emscripten, there is a documented `lto` feature that does the same thing. If emscripten like Clang supports `-flto=thin`/`-flto=full` you might be better to switch the toolset to inherit from Clang instead of GCC.

Reply:
> there is a documented `lto` feature that does the same thing

Should `link-optimization` be renamed to `lto`? The meaning now is the same.

> If emscripten like Clang supports `-flto=thin`/`-flto=full`

No. Despite it's based on Clang Emscripten's LTO implementation is a bit more complicated. See it's [documentation](https://emscripten.org/docs/optimizing/Optimizing-Code.html#lto). In other words `lto-mode` is not applicable here.

In general, renaming to `lto` seems like a right thing because according to the docs above `-flto` should be used during compilation as well.

Reply:
> No. Despite it's based on Clang Emscripten's LTO implementation is a bit more complicated. See its [documentation](https://emscripten.org/docs/optimizing/Optimizing-Code.html#lto). In other words `lto-mode` is not applicable here.

I think https://github.com/emscripten-core/emscripten/issues/10868 tells an opposite thing.

Reply:
Ok, the official Emscripten documentation is out of date. I found [such](https://github.com/emscripten-core/emscripten/blob/f0c4614a279ff57ce331ba3bcb60b076ac171603/emcc.py#L2881) custom processing of `-flto` in emcc.py. Do you think that inheriting from `clang` (which, BTW, on its own inherits from `gcc`) should be done in this PR? I tried to inherit from `clang` but it didn't work out of the box. So, there will be a lot more changes.

Reply:
Did you inherit from `clang` or `clang-linux`? You should have been inheriting from `clang-linux` (that 'linux' suffix means clang with default/posix interface, not the os), sorry I did not tell you that. I would try doing it myself but I don't have the toolset and the CI doesn't tests it either. Don't bother with this if you are short on time.

Reply:
I used `clang-linux`. `clang` is kind of wrapper around `clang-win`, `clang-linux`, etc. I will try to work it out but don't expect too much from me. My knowledge of bjam is rather scarce.